### PR TITLE
Implementation of a monorepo, using Turborepo

### PR DESCRIPTION
**Pull request by tphalp**
Uses https://turbo.build/ to help streamline development of the ucw-app. Resolves UCP-39.

_Note: Most of the 150+ file changes are due to moving files around, and running `lint --fix`_

### Tasks:

- [x] Implement Turborepo folder structure
- [x] Add necessary turborepo files
- [x] Update npm scripts to support turborepo conventions
- [x] Update Docker-related files
- [x] Update GH Action workflows (e2e/unit tests/env)
- [x] Fix lint issues
- [x] Update READMEs
- [x] Create new READMEs for monorepo and docker
- [x] Clean-up docker files for better efficiency
- [x] Update docker hub publish workflow
- [x] Fix husky to work with monorepo structure

### Ways to test:

After pullling this branch (all of this from the root of the project, unless specified): 

- Run `npm run clean`
- Run `npm ci`
- Run `npm run test`. All tests should pass.
- Run `npm run dev`. 
- Once that is running, run `ucw-example` from another terminal, and check the example app in a browser. It should load as expected.
- Kill the above process, but keep the example app running.
- Try running the example app again. It should fail to load the iframe contents.
- Next, run `docker compose up`. 
- Once that finishes, try the example app again. It should load as expected.

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -1,33 +1,6 @@
-CryptoKey="" #generated via a step in the readme
```
Comment (tphalp) [inline .env.example:1]: Split-out the .env file to distinguish between those that are required for docker, vs those that the `server` uses. They used to be all in one .env in the root.

File: .github/workflows/e2e-tests.yml [hunk 2]
```diff
@@ -25,22 +25,21 @@ jobs:
 
       - name: 'Create env file'
```
Comment (tphalp) [inline .github/workflows/e2e-tests.yml:26]: I updated this file to work with the new monorepo folder structure.

File: .github/workflows/unit-tests.yml [hunk 3]
```diff
@@ -31,20 +31,21 @@ jobs:
         env:
```
Comment (tphalp) [inline .github/workflows/unit-tests.yml:31]: Changed the location of the `.env` file that is created, to flow with new monorepo folder structure.

File: DOCKER.md [hunk 4]
```diff
@@ -0,0 +1,27 @@
+# Docker
```
Comment (tphalp) [inline DOCKER.md:1]: New file with docker-specific info.

File: MONOREPO.md [hunk 5]
```diff
@@ -0,0 +1,24 @@
+## This repo uses Turborepo as a monorepo management tool
```
Comment (tphalp) [inline MONOREPO.md:1]: New markdown for monorepo-specific info.

File: apps/server/.env.example [hunk 6]
```diff
@@ -0,0 +1,38 @@
+ # CryptoKey: Generated via a step in the readme
```
Comment (tphalp) [inline apps/server/.env.example:1]: Moved from project root.

File: apps/server/.eslintrc.js [hunk 7]
```diff
@@ -0,0 +1,44 @@
+module.exports = {
```
Comment (tphalp) [inline apps/server/.eslintrc.js:1]: Moved from project root

File: apps/server/README.md [hunk 8]
```diff
@@ -0,0 +1,76 @@
+# Universal Connect Widget (Application)
```
Comment (tphalp) [inline apps/server/README.md:1]: Moved from project root.

File: apps/server/package.json [hunk 9]
```diff
@@ -0,0 +1,98 @@
+{
```
Comment (tphalp) [inline apps/server/package.json:1]: Moved from root.

File: apps/server/src/incubationVcs/vcsService.ts [hunk 10]
```diff
@@ -1,12 +1,12 @@
 import {
```
Comment (tphalp) [inline apps/server/src/incubationVcs/vcsService.ts:1]: File linted.

File: apps/server/src/providers/akoya.ts [hunk 11]
```diff
@@ -1,20 +1,17 @@
 import {
```
Comment (tphalp) [inline apps/server/src/providers/akoya.ts:1]: File linted.

File: apps/server/src/providers/finicity.ts [hunk 12]
```diff
@@ -1,20 +1,17 @@
 import {
```
Comment (tphalp) [inline apps/server/src/providers/finicity.ts:1]: File linted.

File: apps/server/src/providers/index.ts [hunk 13]
```diff
@@ -14,8 +14,8 @@ import type {
   ProviderApiClient,
```
Comment (tphalp) [inline apps/server/src/providers/index.ts:14]: File linted, and paths fixed.

File: apps/server/src/providers/mx.ts [hunk 14]
```diff
@@ -6,11 +6,11 @@ import type {
   Institution,
```
Comment (tphalp) [inline apps/server/src/providers/mx.ts:6]: File linted, and paths fixed.

File: apps/server/src/providers/sophtron.ts [hunk 15]
```diff
@@ -6,19 +6,19 @@ import type {
   Institution,
```
Comment (tphalp) [inline apps/server/src/providers/sophtron.ts:6]: File linted, and paths fixed.

File: apps/server/src/providers/sophtron.ts [hunk 16]
```diff
@@ -6,19 +6,19 @@ import type {
...
+} from '../shared/contract'
+import { mapJobType } from '../utils'
 
 import config from '../config'
 import { debug, trace, error as _error } from '../infra/logger'
 import SophtronClient from '../serviceClients/sophtronClient/v2'
 import SophtronClientV1 from '../serviceClients/sophtronClient'
 
-const uuid = /^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$/
+// const uuid = /^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$/
```
Comment (tphalp) [inline apps/server/src/providers/sophtron.ts:21]: Unused var, per linting.

File: apps/server/src/server.js [hunk 17]
```diff
@@ -83,13 +83,13 @@ if (config.ResourcePrefix !== 'local') {
     }
```
Comment (tphalp) [inline apps/server/src/server.js:83]: File linted, and paths fixed.

File: apps/server/src/shared/connect/fireflyApiBridge/index.js [hunk 18]
```diff
@@ -9,6 +9,7 @@ const api = new ConnectApi({ context })
```
Comment (tphalp) [inline apps/server/src/shared/connect/fireflyApiBridge/index.js:9]: File linted.

File: apps/server/src/shared/contract.ts [hunk 19]
```diff
@@ -1,4 +1,4 @@
-import type { StorageClient } from 'server/serviceClients/storageClient'
+import type { StorageClient } from '../serviceClients/storageClient'
```
Comment (tphalp) [inline apps/server/src/shared/contract.ts:1]: File linted, and paths fixed.

File: apps/server/src/test/handlers.ts [hunk 20]
```diff
@@ -1,5 +1,5 @@
 import { http, HttpResponse } from 'msw'
```
Comment (tphalp) [inline apps/server/src/test/handlers.ts:1]: Paths fixed.

File: apps/server/tsconfig.json [hunk 21]
```diff
@@ -0,0 +1,22 @@
+{
```
Comment (tphalp) [inline apps/server/tsconfig.json:1]: File moved from root.

File: apps/server/tsconfig.paths.json [hunk 22]
```diff
@@ -0,0 +1,9 @@
+{
```
Comment (tphalp) [inline apps/server/tsconfig.paths.json:1]: File moved from root.

File: docker-compose.yaml [hunk 23]
```diff
@@ -12,40 +12,51 @@ services:
     container_name: ucw-app-server
```
Comment (tphalp) [inline docker-compose.yaml:12]: Updated to work with new .env file, and monorepo structure.

File: docker-server.Dockerfile [hunk 24]
```diff
@@ -0,0 +1,58 @@
+# TODO: This comment might need to change, depending on how compose will work with turborepo
```
Comment (tphalp) [inline docker-server.Dockerfile:1]: Move to root. I couldn't get compose to work as expected with Dockerfiles in their respective project folders. This worked nicely.

File: docker-ui.Dockerfile [hunk 25]
```diff
@@ -0,0 +1,51 @@
+# TODO: This comment might need to change, depending on how compose will work with turborepo
```
Comment (tphalp) [inline docker-ui.Dockerfile:1]: Move to root. I couldn't get compose to work as expected with Dockerfiles in their respective project folders. This worked nicely.

File: package.json [hunk 26]
```diff
@@ -1,97 +1,26 @@
 {
```
Comment (tphalp) [inline package.json:1]: New, simplified root package file, to be used with turborepo.

File: README.md [hunk 27]
```diff
@@ -1,58 +1,66 @@
 # Universal Connect Widget (Application)
 
-This repo is a full-stack application which anyone can clone and self-host as a way to serve the connect widget via a url which can then be loaded into an iframe.
+This repo is a full-stack application which anyone can clone and self-host as a way to serve the connect widget via a url, which can then be loaded into an iframe.
 
 ## Getting Started (in production)
 
-To get started, clone the repo, follow the steps in `Initial Setup` (below) to set up your `.env` file, and then run the following command, from the root of the project:
+To get started, clone the repo, follow the steps in [Getting Started](#getting-started-in-development) and  [Initial Setup](#initial-setup) (below) to set up your `.env` files, and then run the following command, from the root of the project:
```
Comment (wesrisenmay-mx) [inline README.md:7]: To get started:

File: apps/server/package.json [hunk 28]
```diff
@@ -0,0 +1,98 @@
...
+    "{.eslintrc.js,.prettierrc.js}": [
+      "prettier --write"
+    ],
+    "*.{js[x],ts[x]}": [
+      "npm run lint -- --cache --fix"
+    ]
+  },
+  "husky": {
+    "hooks": {
+      "pre-commit": "prettier --write . && lint-staged && npm run test && git add -A ."
```
Comment (wesrisenmay-mx) [inline apps/server/package.json:95]: Are these hooks working? I would guess that they need to be in the root package.json to work

Comment (tphalp) [reply apps/server/package.json:95]: Good catch. I'll take a look.

File: README.md [hunk 29]
```diff
@@ -1,58 +1,66 @@
 # Universal Connect Widget (Application)
 
-This repo is a full-stack application which anyone can clone and self-host as a way to serve the connect widget via a url which can then be loaded into an iframe.
+This repo is a full-stack application which anyone can clone and self-host as a way to serve the connect widget via a url, which can then be loaded into an iframe.
```
[the remark this answers is not in the document]
Comment (tphalp) [reply README.md]: I read that pdf in more detail, and I would argue the comma does belong. I think it falls under the "Introductory bits" section. What do you think?

File: .husky/pre-commit [hunk 30]
```diff
@@ -1 +1,2 @@
-# npm run lint:fix TODO: uncomment this when ts errors are all fixed
+npx lint-staged
+npm run test
```
Comment (wesrisenmay-mx) [inline .husky/pre-commit:2]: I don't think we should run tests on commit. This works for now while we have low test coverage, but eventually these tests will take too long to run for how frequently we commit. Our pipeline will protect us from merging things that break tests.

Comment (tphalp) [reply .husky/pre-commit]: Removed

**Review by wesrisenmay-mx — APPROVED** [review]

**Pull request merged**